api/checks/update.go

```diff
@@ -0,0 +1,150 @@
+// Copyright 2018 The WPT Dashboard Project. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package checks
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/deckarep/golang-set"
+
+	"github.com/gorilla/mux"
+	"github.com/web-platform-tests/wpt.fyi/api/checks/summaries"
+	"github.com/web-platform-tests/wpt.fyi/shared"
+	"google.golang.org/appengine"
+)
+
+// updateCheckHandler handles /api/checks/[commit] POST requests.
+func updateCheckHandler(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	sha := vars["commit"]
+	if len(sha) != 40 {
+		http.Error(w, fmt.Sprintf("Invalid commit: %s", sha), http.StatusBadRequest)
+		return
+	}
+
+	ctx := appengine.NewContext(r)
+	filter, err := shared.ParseTestRunFilterParams(r)
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	} else if len(filter.Products) < 1 {
+		http.Error(w, "product param is missing", http.StatusBadRequest)
+		return
+	}
+	filter.SHA = sha[:10]
+	one := 1
+	runs, err := shared.LoadTestRuns(ctx, filter.Products, filter.Labels, sha[:10], filter.From, filter.To, &one, nil)
+	allRuns := runs.AllRuns()
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	} else if len(allRuns) < 1 {
+		http.NotFound(w, r)
+		return
+	} else if len(allRuns) > 1 {
+		http.Error(w, fmt.Sprintf("Expected exactly 1 run, but found %v", len(runs)), http.StatusBadRequest)
+		return
+	}
+
+	// Get a master run to compare.
+	labels := filter.Labels
+	if labels == nil {
+		labels = mapset.NewSet()
+	}
+	labels.Add("master")
+	masterRuns, err := shared.LoadTestRuns(ctx, filter.Products, labels, "", nil, nil, nil, nil)
+	allMasterRuns := masterRuns.AllRuns()
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	} else if len(allMasterRuns) < 1 {
+		http.Error(w, "No master run found to compare differences", http.StatusNotFound)
+		return
+	}
+
+	beforeJSON, err := shared.FetchRunResultsJSON(ctx, r, allRuns[0])
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Failed to fetch 'before' results: %s", err.Error()), http.StatusInternalServerError)
+		return
+	}
+	afterJSON, err := shared.FetchRunResultsJSON(ctx, r, allMasterRuns[0])
+	if err != nil {
+		http.Error(w, fmt.Sprintf("Failed to fetch 'after' results: %s", err.Error()), http.StatusInternalServerError)
+		return
+	}
+
+	product, _ := shared.ParseProductSpec(runs[0].Product.String())
+	checkState := summaries.CheckState{
+		Product:    product,
+		HeadSHA:    sha,
+		Title:      getCheckTitle(product),
+		DetailsURL: getMasterDiffURL(ctx, sha, product),
+		Status:     "in_progress",
```

Comment:
@lukebjerring stumbled upon this again and I'm wondering shouldn't this be "completed"?